In [4]:
import numpy as np
import matplotlib.pyplot as plt

plt.style.use('../mystyle.mplstyle')

import sys
import os

# Get the current directory
current_dir = os.getcwd()

# Go one level up to the parent directory
parent_dir = os.path.abspath(os.path.join(current_dir, os.pardir))

# Add the parent directory to the system path
sys.path.append(parent_dir)

# Now you can import the module
from pyfiles.image_tools import trim_whitespace

`````{admonition} Remark
:class: important

Please be aware that these lecture notes are accessible online in an '**early access**' format. They are actively being developed, and certain sections will be further enriched to provide a comprehensive understanding of the subject matter.

`````

# Basic Definitions and Concepts

## Ellipsoid

An Earth ellipsoid, also known as an Earth spheroid, is a mathematical model that closely resembles the Earth's shape. It is characterized by:

1. A spheroid (ellipsoid of revolution) shape
2. A minor axis connecting the geographical North and South Poles
3. Alignment with the Earth's axis of rotation

The ellipsoid is defined by two main parameters:

1. **Equatorial axis (a)**: The longer radius at the equator
2. **Polar axis (b)**: The shorter radius from the center to either pole

The difference between these axes is approximately 21 km, or 0.335% of the equatorial axis.


{numref}`Ellipsoid_revolution_oblate_aab_auxiliary_sphere` illustrates an Earth ellipsoid, specifically an oblate spheroid. The longer equatorial axis (a) extends horizontally, while the shorter polar axis (b) extends vertically. The dashed line represents an auxiliary sphere with a radius equal to the equatorial axis, highlighting the difference between a perfect sphere and an ellipsoid. This visualization helps in understanding how the Earth's shape deviates from a perfect sphere, being slightly flattened at the poles and bulging at the equator.

```{figure} Ellipsoid_revolution_oblate_aab_auxiliary_sphere.png
---
width: 400px
align: center
name: Ellipsoid_revolution_oblate_aab_auxiliary_sphere
---
Visual representation of an Earth ellipsoid. Credit: [Wikipedia.org](https://en.wikipedia.org/wiki/Geographic_coordinate_system). [Link to Image](https://upload.wikimedia.org/wikipedia/commons/8/8c/Ellipsoid_revolution_oblate_aab_auxiliary_sphere.svg).
```

## The Geoid

The geoid is a complex and highly detailed model that represents the Earth's true shape more accurately than a simple ellipsoid. Defined as an equipotential surface of the Earth's gravity field, the geoid closely approximates mean sea level. Essentially, it reflects the shape that the ocean surface would adopt under the influence of Earth's gravitational pull and rotation, without the interference of other factors such as winds and tides.

The color-coded map in {numref}`figure_Earth_Gravitational_Model_1996` illustrates the undulation of the geoid in meters, based on the [EGM96 gravity model](https://cddis.nasa.gov/926/egm96/egm96.html) and the WGS84 reference ellipsoid. The color spectrum ranges from deep blue to red, indicating variations in geoid height from approximately -100 to +80 meters. This map employs a rectangular projection with latitude lines spanning from 75° North to 75° South and longitude lines from 180° West to 180° East. The color variations represent the irregular shape of Earth's gravity field, which deviates from a perfect ellipsoid due to uneven mass distribution within the planet.

```{figure} Earth_Gravitational_Model_1996.png
---
width: 600px
align: center
name: figure_Earth_Gravitational_Model_1996
---
Color-coded map showing variations in Earth's gravitational field, with geoid heights measured in meters according to EGM96 data overlaid on the WGS84 reference ellipsoid. Credit: [Wikipedia.org](https://en.wikipedia.org/wiki/Geoid). [Link to Image](https://en.wikipedia.org/wiki/Geoid#/media/File:Earth_Gravitational_Model_1996.png).
```

The geoid exhibits several important characteristics {cite}`kumar2023geographic,freymueller2023geodesy`:

1. **Irregular Shape**: Unlike the smooth mathematical ellipsoid, the geoid has an uneven, wavy surface due to the Earth's varying mass distribution and gravitational anomalies.

2. **Gravitational Equipotential Surface**: It represents the shape that the ocean surface would take under the influence of Earth's gravity and rotation alone. All points on the geoid have the same gravitational potential, making it a critical reference for understanding gravitational effects.

3. **Reference for Elevation**: The geoid serves as a crucial reference surface for measuring orthometric heights, which are commonly understood as heights above mean sea level.

4. **Relationship to Mean Sea Level**: The geoid closely approximates mean sea level, extending across the continents and providing a realistic baseline for elevation measurements.

5. **Geoid Undulation**: The difference between the geoid and the reference ellipsoid is termed geoid undulation or geoidal height. This undulation can vary significantly, with differences of up to 110 meters observed in various locations.

6. **Gravity Anomalies**: The shape of the geoid reflects local variations in the Earth's gravity field, with elevations in areas of higher density and depressions in regions of lower density.

{numref}`Geoid_undulation_10k_scale` shows a color-coded representation of the geoid, which represents mean sea level across the globe. The colors range from red to blue, indicating variations in gravitational strength and corresponding geoid heights. Red areas signify higher elevations of the geoid, while blue areas indicate lower elevations. This model helps in understanding global ocean circulation, sea level changes, and earth sciences such as geodesy and geography. It visualizes the irregular shape of Earth’s gravity field rather than its actual surface or a perfect sphere.

```{figure} Geoid_undulation_10k_scale.jpg
---
width: 400px
align: center
name: Geoid_undulation_10k_scale
---
Geoid Representation: Variations in Earth’s Gravitational Field. Credit: [Wikipedia.org](https://en.wikipedia.org/wiki/Geoid). [Link to Image](https://upload.wikimedia.org/wikipedia/commons/4/4a/Geoid_undulation_10k_scale.jpg).
```

Understanding the geoid is essential for several reasons {cite}`seeber2008satellite,sjöberg2017gravity`:

1. **Accurate Elevation Measurements**: The geoid provides a more precise reference for determining elevations compared to a simple ellipsoid model.

2. **GPS Height Conversion**: It plays a vital role in converting GPS-derived ellipsoidal heights into practical orthometric heights, which are more relevant for everyday applications.

3. **Earth Science Applications**: The geoid is critical for various fields, including oceanography, hydrology, and geology, as it influences water flow and sea level studies.

4. **Improved Mapping and Surveying**: The geoid enables more accurate cartography, digital terrain models, and satellite altimetry, enhancing the quality of geographic information systems.

5. **Gravity Field Studies**: It aids in the study of the Earth's gravity field and its variations over time, contributing to our understanding of geophysical processes.

Determining the precise shape of the geoid remains a significant challenge in geodesy. Modern techniques combine satellite data from missions like [GRACE](https://earthobservatory.nasa.gov/features/GRACE/page1.php) and [GOCE](https://earth.esa.int/eogateway/missions/goce) with terrestrial gravity measurements to create increasingly accurate geoid models. These advancements enhance our understanding of the Earth's shape, gravity field, and various geophysical processes, driving progress in both scientific research and practical applications {cite}`szelachowska_contribution_2022, tapley_contributions_2019, van_der_meijde_goce_2015`.

## Datum

A datum is a reference system that defines the size and shape of the Earth and provides a foundation for calculating the positions of points on the Earth's surface. It integrates the concepts of the ellipsoid and geoid to create a practical framework for mapping and spatial measurements {cite}`hofmann2006physical, seeber_satellite_2008`.

### Key Components of a Datum:

1. **Reference Ellipsoid**: A mathematically defined surface that approximates the Earth's shape {cite}`torge_geodesy_2012`.
2. **Origin Point**: A specific point where the ellipsoid is anchored to the Earth's surface {cite}`seeber_satellite_2008`.
3. **Orientation**: The alignment of the ellipsoid's axes relative to the Earth {cite}`hofmann2006physical`.

### Relationship to Ellipsoid and Geoid:

- **Ellipsoid Connection**: The datum specifies the ellipsoid to be used and its position relative to the Earth's center and surface {cite}`torge_geodesy_2012`.
- **Geoid Consideration**: Although based on an ellipsoid, a datum often incorporates geoid undulations for more accurate vertical measurements {cite}`hofmann2006physical`.
- **Datum's Role**: A datum integrates both the ellipsoid and geoid by defining the ellipsoid's position relative to the geoid and the Earth's surface, allowing for accurate three-dimensional positioning by combining horizontal and vertical data {cite}`seeber_satellite_2008`.

### Types of Datums:

1. **Geodetic Datums**:
   - **Purpose**: To accurately describe locations in terms of latitude and longitude for horizontal positioning on the Earth's surface.
   - **Basis**: An ellipsoid that approximates the Earth's shape, considering its equatorial bulge and polar flattening {cite}`torge_geodesy_2012`.
   - **Components**: Ellipsoid specifications, position and orientation relative to the Earth, and the datum origin.
   - **Examples**:
     - **Geocentric Datums**:
       - Origin at the Earth's center of mass {cite}`seeber_satellite_2008`.
       - <font color='Blue'><b>Example:</b></font> [WGS84](https://epsg.io/4326) (World Geodetic System 1984)
       - Used for global applications like GPS.
     - **Local Datums**:
       - Designed to fit a specific region of the Earth's surface.
       - <font color='Blue'><b>Example:</b></font> [NAD83](https://epsg.io/4269) (North American Datum 1983)
       - Provides enhanced accuracy for localized regions {cite}`hofmann2006physical`.

2. **Vertical Datums**:
   - **Purpose**: To establish a reference surface for elevation measurements, used for determining heights or depths.
   - **Basis**: Typically based on mean sea level or a geoid model {cite}`torge_geodesy_2012`.
   - **Key Feature**: Defines a "zero point" from which heights are measured.
   - **Examples**:
     - NAVD88 (North American Vertical Datum of 1988)
     - EGM96 (Earth Gravitational Model 1996) {cite}`seeber_satellite_2008`.

### Importance in GIS and Mapping:

- **Coordinate Reference**: Datums provide the foundational framework for coordinate systems, enabling precise location representation {cite}`hofmann2006physical`.
- **Consistency**: Ensure that all users of a particular system reference the same spatial framework {cite}`torge_geodesy_2012`.
- **Transformation**: Facilitate conversions between different coordinate systems and datums {cite}`seeber_satellite_2008`.
- **Data Integration**: Allow for the accurate combination of spatial data from various sources {cite}`hofmann2006physical`.
- **Spatial Accuracy**: Ensure that coordinates accurately reflect real-world locations {cite}`torge_geodesy_2012`.
- **Analysis Reliability**: Support accurate measurements, distances, and area calculations {cite}`seeber_satellite_2008`.
- **Global Compatibility**: Enable the use of global navigation and positioning systems {cite}`hofmann2006physical`.

### Practical Implications:

- Changing datums can significantly alter the coordinate values for the same physical location {cite}`torge_geodesy_2012`.
- Understanding datums is crucial for accurate data integration and analysis in GIS {cite}`seeber_satellite_2008`.
- Although modern GIS software often handles datum transformations automatically, users must understand the underlying concepts to ensure accuracy {cite}`hofmann2006physical`.
- Datums are essential for applications including mapping, navigation, urban planning, environmental monitoring, and scientific research {cite}`torge_geodesy_2012`.

In [5]:
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature

# Create a figure with two subplots
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(20, 10), subplot_kw={'projection': ccrs.PlateCarree()})

# WGS84 Map
ax1.set_title('WGS84 (EPSG:4326)', fontsize=16)
ax1.add_feature(cfeature.LAND)
ax1.add_feature(cfeature.OCEAN)
ax1.add_feature(cfeature.COASTLINE)
ax1.add_feature(cfeature.BORDERS, linestyle=':')
ax1.gridlines(draw_labels=True)
ax1.set_global()

# NAD83 Map
ax2.set_title('NAD83 (EPSG:4269)', fontsize=16)
ax2.add_feature(cfeature.LAND)
ax2.add_feature(cfeature.OCEAN)
ax2.add_feature(cfeature.COASTLINE)
ax2.add_feature(cfeature.BORDERS, linestyle=':')
ax2.gridlines(draw_labels=True)
ax2.set_extent([-170, -50, 15, 75], crs=ccrs.PlateCarree())  # North America focus

plt.tight_layout()
plt.close()

## Meridians and Parallels

Meridians and parallels form the grid system used in geographic coordinate systems to precisely locate any point on Earth's surface.

- **Meridians:** Meridians are imaginary vertical lines that run from the North Pole to the South Pole. They are used to measure longitude, which indicates how far east or west a location is from the Prime Meridian.

    - Meridians are half-circles, with each meridian having an anti-meridian on the opposite side of the globe.
    - They converge at the poles and are farthest apart at the equator.
    - Longitude is measured in degrees, from 0° to 180° east or west of the Prime Meridian.

    **Prime Meridian**: 
    - Located at 0° longitude, passing through Greenwich, England (shown as the red line in {numref}`3d_map_meridians`).
    - Serves as the reference point for measuring longitude east and west.
    <!-- - The 180° meridian, opposite the Prime Meridian, is known as the International Date Line (shown as the blue line in {numref}`3d_map_meridians`). -->

    ```{figure} 3d_map_meridians.png
    ---
    width: 500px
    align: center
    name: 3d_map_meridians
    ---
    Meridians and Prime Meridian.
    ```


- **Parallels:** Parallels are imaginary horizontal lines that circle the globe parallel to the equator. They are used to measure latitude, which indicates how far north or south a location is from the equator.

    - Parallels are complete circles that gradually decrease in size from the equator towards the poles.
    - They are always equidistant from each other.
    - Latitude is measured in degrees, from 0° at the equator to 90° at the poles (north or south).

    **Equator**:
    - The largest parallel, located at 0° latitude (shown as the red line in {numref}`3d_map_parallels`).
    - Divides the Earth into the Northern and Southern Hemispheres.
    - Has a constant distance of approximately 40,075 km (24,901 miles) around the globe.

    **Other Notable Parallels**:
    - Tropic of Cancer (23.5°N) and Tropic of Capricorn (23.5°S): Mark the northernmost and southernmost latitudes where the Sun can appear directly overhead.
    - Arctic Circle (66.5°N) and Antarctic Circle (66.5°S): Mark the southernmost and northernmost latitudes where 24-hour daylight or darkness can occur (shown as the blue lines in {numref}`3d_map_parallels`).

    ```{figure} 3d_map_parallels.png
    ---
    width: 500px
    align: center
    name: 3d_map_parallels
    ---
    Meridians, Prime Meridian, and International Date Line.
    ```

In [6]:
# 3d_map_meridians
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature
from matplotlib.lines import Line2D

# Create figure and set up projection
fig, ax = plt.subplots(figsize=(6, 6),
                       subplot_kw={'projection': ccrs.Orthographic(10, 10)})

# Draw meridians
for lon in range(-180, 181, 30):
    gl = ax.gridlines(xlocs=[lon], color='black', linestyle='--', linewidth=1.5)
    gl.ylines = False  # Turn off latitude lines

# Highlight the Prime Meridian and International Date Line
gl = ax.gridlines(xlocs=[0], color='red', linestyle='-', linewidth=2)
gl.ylines = False  # Turn off latitude lines

# gl = ax.gridlines(xlocs=[180, -180], color='blue', linestyle='-', linewidth=2)
# gl.ylines = False  # Turn off latitude lines

# Set the map extent
ax.set_global()
ax.stock_img()
ax.coastlines()

# Add a title
# plt.title("Meridians, Prime Meridian, and International Date Line", fontsize=14, weight='bold')

# Add a legend
legend_elements = [
    Line2D([0], [0], color='black', linestyle='--', lw = 1.5, label='Meridians'),
    Line2D([0], [0], color='red', lw = 2, label='Prime Meridian'),
    # Line2D([0], [0], color='blue', lw = 2,  label='International Date Line')
]
fig.legend(handles=legend_elements, loc='lower right', fontsize=14, ncol=1)

# # Add a watermark
# ax.text(0.05, 0.05, '@Dastour', transform=ax.transAxes,
#         fontsize=14, color='k', alpha=1,
#         ha='center', va='center')

plt.tight_layout()
fig.savefig('3d_map_meridians.png', bbox_inches='tight', dpi=300)
plt.close()

In [7]:
# 3d_map_parallels
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature
from matplotlib.lines import Line2D

# Create figure and set up projection
fig, ax = plt.subplots(figsize=(6, 6),
                       subplot_kw={'projection': ccrs.Orthographic(10, 10)})

# Draw parallels
for lat in range(-90, 91, 30):
    gl = ax.gridlines(ylocs=[lat], color='black', linestyle='--', linewidth=1.5)
    gl.xlines = False  # Turn off latitude lines

# Highlight the Equator and notable parallels
gl = ax.gridlines(ylocs=[0], color='red', linestyle='-', linewidth=2)
gl.xlines = False  # Turn off latitude lines
gl = ax.gridlines(ylocs=[23.5, -23.5, 66.5, -66.5], color='blue', linestyle='-', linewidth=2)
gl.xlines = False  # Turn off latitude lines

# Set the map extent
ax.set_global()
ax.stock_img()
ax.coastlines()

# Add a title
# plt.title("Parallels, Equator, and Notable Parallels", fontsize=14, weight='bold')

# Add a legend
legend_elements = [
    Line2D([0], [0], color='black', linestyle='--', lw = 1.5, label='Parallels'),
    Line2D([0], [0], color='red', lw = 2, label='Equator'),
    Line2D([0], [0], color='blue', lw = 2, label='Notable Parallels')
]
fig.legend(handles=legend_elements, loc='lower right', fontsize=14, ncol=1)

plt.tight_layout()
fig.savefig('3d_map_parallels.png', bbox_inches='tight', dpi=300)
plt.close()

## Latitude ($\phi$) and Longitude ($\lambda$)

In the geographic coordinate system, locations on Earth are specified using two angular measurements: latitude ($\phi$) and longitude ($\lambda$).

- **Latitude ($\phi$):** Latitude, denoted by the Greek letter phi ($\phi$), measures the angular distance north or south of the Earth's equator. It is expressed in degrees, ranging from $-90°$ to $+90°$.

  - $\phi = 0°$ represents the equator
  - $\phi = +90°$ corresponds to the North Pole
  - $\phi = -90°$ corresponds to the South Pole

  Lines of constant latitude are called parallels. They form circles on the Earth's surface that are parallel to the equator.

  - Positive values ($\phi > 0°$) indicate locations in the **Northern Hemisphere** (N).
  - Negative values ($\phi < 0°$) indicate locations in the **Southern Hemisphere** (S).

- **Longitude ($\lambda$)**: Longitude, denoted by the Greek letter lambda ($\lambda$), measures the angular distance east or west of the prime meridian. It is expressed in degrees, ranging from $-180°$ to $+180°$.

  - $\lambda = 0°$ represents the prime meridian (passing through Greenwich, UK)
  - $\lambda = +180°$ and $\lambda = -180°$ represent the International Date Line

  Lines of constant longitude are called meridians. They form great circles that converge at the poles.

  - Positive values ($\lambda > 0°$) indicate locations in the **Eastern Hemisphere** (E).
  - Negative values ($\lambda < 0°$) indicate locations in the **Western Hemisphere** (W).



{numref}`lon_lat` visually explains the concepts of **latitude** ($\phi$) and **longitude** ($\lambda$) in the geographic coordinate system.

<!-- - **Latitude ($\phi$)**: The left circle shows how latitude is measured as the angular distance from the equator. The angle $\phi$ is marked from the center of the Earth to a point on the surface, intersecting the equator. This demonstrates how latitude ranges from $-90°$ (South Pole) to $+90°$ (North Pole).

- **Longitude ($\lambda$)**: The right circle illustrates longitude as the angular distance from the prime meridian. The angle $\lambda$ is marked from the center of the Earth to a point on the surface, intersecting both the equator and the prime meridian. Longitude ranges from $-180°$ (west) to $+180°$ (east). -->

```{figure} lon_lat.png
---
width: 400px
align: center
name: lon_lat
---
Visualization of latitude ($\phi$) and longitude ($\lambda$).
```

In [8]:
# import matplotlib.pyplot as plt
# import numpy as np
# from matplotlib.patches import Arc

# # Create a figure and axes
# fig, ax = plt.subplots(1, 2, figsize=(9.5, 5))
# center = (0, 0)
# radii = np.linalg.norm((1, 1)) / 2

# # Function to draw the common elements of the plots
# def draw_plot(ax, point, angle_label, angle_start, angle_end, point_label):
#     # Draw circle representing the sphere
#     circle = plt.Circle(center, 1, edgecolor='black', facecolor='none', lw=2)
#     ax.add_patch(circle)

#     # Draw equator line
#     equator_line = ax.hlines(center[0], -1., 1., colors='blue', linestyles='solid', lw=2, label='Equator')

#     # Draw prime meridian line
#     prime_meridian_line = ax.vlines(center[1], -1., 1., colors='red', linestyles='dashed', lw=2, label='Prime Meridian')

#     # Draw point and line to center
#     ax.plot(point[0], point[1], 'ko')  # Point
#     ax.plot([point[0], center[0]], [point[1], center[1]], 'k-')  # Line from point to center
#     ax.text(point[0] + 0.1, point[1], point_label, ha='center', fontsize=14)

#     # Draw angle
#     arc = Arc(center, 0.4, 0.4, angle=0, theta1=angle_start, theta2=angle_end, color='black')
#     ax.add_patch(arc)
#     ax.text(0.1, 0.3, angle_label, ha='center', fontsize=14)

#     # Set limits and aspect
#     ax.set_xlim(-1.2, 1.2)
#     ax.set_ylim(-1.2, 1.2)
#     ax.set_aspect('equal')
#     ax.axis('off')

#     return equator_line, prime_meridian_line

# # Draw the left plot
# equator_line, prime_meridian_line = draw_plot(ax[0], (radii, radii), r'$\lambda$', 45, 90, 'x')

# # Draw the right plot
# draw_plot(ax[1], (radii, radii), r'$\varphi$', 0, 45, 'y')

# ax[0].set_title(r'(a) Longitude ($\lambda$)')
# ax[1].set_title(r'(a) Latitude ($\varphi$)')
# # Create a single legend for the entire figure
# fig.legend(handles=[equator_line, prime_meridian_line], loc='lower center', fontsize=14, ncol=1)

# # Show the plot
# plt.tight_layout()
# fig.savefig('lon_lat.png', bbox_inches='tight', dpi=300)

In [9]:
# lon_lat
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from matplotlib.lines import Line2D
from matplotlib.patches import Patch

# Create a figure and 3D axis
fig = plt.figure(figsize=(8, 8))
ax = fig.add_subplot(111, projection='3d')

# Create a sphere
u = np.linspace(0, 2 * np.pi, 100)
v = np.linspace(0, np.pi, 100)
x = np.outer(np.cos(u), np.sin(v))
y = np.outer(np.sin(u), np.sin(v))
z = np.outer(np.ones(np.size(u)), np.cos(v))
ax.plot_surface(x, y, z, color='blue', edgecolor='none', alpha=0.1)

# Define point P on the sphere
theta = np.pi / 4  # 45 degrees for example
P = (np.cos(theta), np.cos(theta), np.sin(theta))

# Line from center to P
ax.plot([0, P[0]], [0, P[1]], [0, P[2]], 'k-', lw=2)

# Prime meridian (x-z plane)
ax.plot(np.sin(u), 0, np.cos(u), color='k', linestyle='dashed', lw=2, label='Prime Meridian')

# Equator (x-y plane)
ax.plot(np.cos(u), np.sin(u), 0, color='blue', lw=2, label='Equator')

# Angle annotations and shading

# Angle phi: between the black line and the xy-plane
phi_plane = [[(0, 0, 0), (P[0], P[1], 0), (P[0], P[1], P[2])]]
ax.add_collection3d(Poly3DCollection(phi_plane, color='green', alpha=0.2))

ax.plot([P[0], P[0]], [P[1], P[1]], [0, P[2]], color='green', linestyle='--')
ax.text(P[0]/2, P[1]/2, P[2]/2+0.05, r'$\phi$', ha='center', fontsize=14, color='green')

# Projection of point P onto xy-plane
P_proj_xy = (P[0], P[1], 0)

# Angle lambda: between the projection of the black line onto the xy-plane and the xz-plane
lambda_plane = [[(0, 0, 0), (P_proj_xy[0], P_proj_xy[1], 0), (P[0], 0, 0)]]
ax.add_collection3d(Poly3DCollection(lambda_plane, color='OrangeRed', alpha=0.3))

ax.plot([P_proj_xy[0], P_proj_xy[0]], [P_proj_xy[1], 0], [0, 0], color='OrangeRed', linestyle='--')
ax.text(P_proj_xy[0]/2, P_proj_xy[1]/2 - 0.2, 0, r'$\lambda$', ha='center', fontsize=14, color='OrangeRed')

# Set labels
# Remove axis labels and ticks
ax.set_xticks([])
ax.set_yticks([])
ax.set_zticks([])
ax.set_xlabel('')
ax.set_ylabel('')
ax.set_zlabel('')

# Set limits and aspect
ax.set_xlim([-1.2, 1.2])
ax.set_ylim([-1.2, 1.2])
ax.set_zlim([-1.2, 1.2])
ax.set_box_aspect([1, 1, 1])

# Turn off the axis to remove the box
ax.axis('off')

# Create custom legend handles with rectangular shapes
legend_elements = [
    Patch(facecolor='green', edgecolor='green', alpha=0.5, label=r'$\phi$ Angle'),
    Patch(facecolor='OrangeRed', edgecolor='OrangeRed', alpha=0.5, label=r'$\lambda$ Angle'),
    Line2D([0], [0], color='k', linestyle='dashed', lw=2, label='Prime Meridian'),
    Line2D([0], [0], color='blue', lw=2, label='Equator')
]
# Add legend to the plot
fig.legend(handles=legend_elements, loc='lower center', fontsize=12, bbox_to_anchor=(.5, 0.15), ncols = 2)
ax.grid(False)

# Set the view angle: elevation and azimuthal angle
ax.view_init(elev= 15, azim= 2)

# Show the plot
plt.tight_layout()
fig.savefig('lon_lat.png', bbox_inches='tight', dpi=300)
plt.close()

trim_whitespace('lon_lat.png')

'lon_lat.png'

````{prf:example}
:label: example2.2.1

{numref}`3d_map_calgary` shows a world map in an orthographic projection with Calgary at the center, highlighting its global position with coordinates $\phi \approx 51.0447°$N, $\lambda \approx 114.0719°$W.

```{figure} 3d_map_calgary.png
---
width: 600px
align: center
name: 3d_map_calgary
---
World Map Centered on Calgary. 
```
````

In [10]:
# 3d_map_calgary
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature

# Coordinates for Calgary
calgary_lat, calgary_lon = 51.0447, -114.0719

import numpy as np

# Define constants
EARTH_RADIUS_KM = 6371  # Average radius of Earth in kilometers

def deg_to_rad(deg):
    return np.deg2rad(deg)

def rad_to_deg(rad):
    return np.rad2deg(rad)

def dms_to_dd(degrees, minutes, seconds):
    return degrees + minutes/60 + seconds/3600

def dd_to_dms(dd):
    degrees = int(dd)
    minutes = int((dd - degrees) * 60)
    seconds = (dd - degrees - minutes/60) * 3600
    return degrees, minutes, seconds

def haversine_distance(lat1, lon1, lat2, lon2):
    lat1, lon1, lat2, lon2 = map(deg_to_rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arctan2(np.sqrt(a), np.sqrt(1-a))
    return EARTH_RADIUS_KM * c

print(f"Calgary's coordinates: {calgary_lat:.4f}°N, {abs(calgary_lon):.4f}°W")

# Convert to DMS
lat_d, lat_m, lat_s = dd_to_dms(calgary_lat)
lon_d, lon_m, lon_s = dd_to_dms(abs(calgary_lon))
print(f"Calgary's coordinates (DMS): {lat_d}°{lat_m}'{lat_s:.2f}\"N, {lon_d}°{lon_m}'{lon_s:.2f}\"W")


# Create figure and set up projection
fig, ax = plt.subplots(figsize=(6, 6),
                       subplot_kw={'projection': ccrs.Orthographic(calgary_lon, calgary_lat)})

# Add map features
# ax.stock_img()
# ax.coastlines()
ax.add_feature(cfeature.LAND)
ax.add_feature(cfeature.OCEAN)
ax.add_feature(cfeature.COASTLINE)
ax.add_feature(cfeature.BORDERS, linestyle='--')
ax.add_feature(cfeature.LAKES)
ax.add_feature(cfeature.RIVERS)

# Gridlines with larger font size
ax.gridlines(draw_labels=True, dms=True, x_inline=True, y_inline=True,
             xlabel_style={'size': 12}, ylabel_style={'size': 12})

# Plot and annotate Calgary
ax.plot(calgary_lon, calgary_lat, 'ro', markersize=5, transform=ccrs.Geodetic())
ax.text(calgary_lon, calgary_lat + 3, f'Calgary\n({calgary_lat:.4f}°N, {calgary_lon:.4f}°W)', 
        color='red', fontweight='normal', fontsize=12, transform=ccrs.Geodetic(),
        va='bottom', ha='center', bbox=dict(facecolor='white', alpha=0.7, edgecolor='none'))

# Final adjustments
ax.set_global()
# ax.set_title('Orthographic Projection Centered on Calgary', fontsize=14, weight='bold')
plt.tight_layout()
fig.savefig('3d_map_calgary.png', bbox_inches='tight', dpi=300)
plt.close()

Calgary's coordinates: 51.0447°N, 114.0719°W
Calgary's coordinates (DMS): 51°2'40.92"N, 114°4'18.84"W


## Degrees, Minutes, and Seconds (DMS)

Degrees, Minutes, and Seconds (DMS) is a traditional format for expressing geographic coordinates, commonly used in navigation and mapping. This system divides a degree into smaller units to provide a more precise location on the Earth's surface.

**Structure of DMS:**

- **Degrees (°)**: The largest unit, representing the primary division of a circle. There are 360 degrees in a full circle.
- **Minutes (')**: Each degree is divided into 60 minutes.
- **Seconds (")**: Each minute is further divided into 60 seconds.

The relationship can be summarized as:  
\begin{equation}1° = 60' = 3600''\end{equation}

**Common Formats:**

1. **DMS Format**: Expressed as degrees, minutes, and seconds (DD° MM' SS.S").
2. **Degrees and Decimal Minutes (DMM)**: Expressed as degrees and decimal minutes (DD° MM.MMM').
3. **Decimal Degrees (DD)**: Expressed as decimal degrees (DD.DDDDDD°).

**Conversion Between Formats:**

1. **To Convert DMS to Decimal Degrees (DD):**

   \begin{equation}
   \text{DD} = \text{Degrees} + \left(\dfrac{\text{Minutes}}{60}\right) + \left(\dfrac{\text{Seconds}}{3600}\right)
   \end{equation}

   <font color='Blue'><b>Example:</b></font> Given: 51° 2' 40.92"  
   \begin{equation*}
   \text{DD} = 51 + \left(\dfrac{2}{60}\right) + \left(\dfrac{40.92}{3600}\right) \approx 51.0447°
   \end{equation*}

2. **To Convert Decimal Degrees (DD) to DMS:**

   - **Step 1**: The whole number part is the degrees.
   - **Step 2**: Multiply the decimal part by 60 to get minutes:
     \begin{equation}
     \text{Minutes} = (\text{DD} - \text{Degrees}) \times 60
     \end{equation}
   - **Step 3**: Multiply the remaining decimal part by 60 to get seconds:
   
     \begin{equation}
     \text{Seconds} = (\text{Remaining Decimal}) \times 60
     \end{equation}

   <font color='Blue'><b>Example:</b></font> Given: 51.0447°  
   \begin{equation*}
   \text{Degrees} = 51°
   \end{equation*}
   \begin{equation*}
   \text{Minutes} = (0.0447) \times 60 \approx 2.682'
   \end{equation*}
   \begin{equation*}
   \text{Seconds} = (0.682) \times 60 \approx 40.92''
   \end{equation*}
   Result: 51° 2' 40.92"

3. **To Convert Decimal Degrees (DD) to Degrees and Decimal Minutes (DMM):**

   - **Step 1**: The whole number part is the degrees.
   - **Step 2**: Multiply the decimal part by 60 to get decimal minutes:
     \begin{equation}
     \text{Decimal Minutes} = (\text{DD} - \text{Degrees}) \times 60
     \end{equation}

   <font color='Blue'><b>Example:</b></font> Given: 51.0447°  
   \begin{equation*}
   \text{Degrees} = 51°
   \end{equation*}

   \begin{equation*}
   \text{Decimal Minutes} = (0.0447) \times 60 \approx 2.682'
   \end{equation*}
   Result: 51° 2.682'

**Usage Considerations:**

While DMS is still prevalent in many contexts, particularly in traditional navigation, modern applications often prefer Decimal Degrees due to their ease of use in calculations and compatibility with digital systems.

In [11]:
import math

def dd_to_dms(dd):
    """Convert decimal degrees to degrees, minutes, seconds"""
    degrees = int(dd)
    minutes = int((dd - degrees) * 60)
    seconds = (dd - degrees - minutes/60) * 3600
    return degrees, minutes, seconds

def dms_to_dd(degrees, minutes, seconds):
    """Convert degrees, minutes, seconds to decimal degrees"""
    return degrees + (minutes / 60.0) + (seconds / 3600.0)

def dd_to_dmm(dd):
    """Convert decimal degrees to degrees and decimal minutes"""
    degrees = int(dd)
    minutes = (dd - degrees) * 60
    return degrees, minutes

# Calgary's coordinates
calgary_lat, calgary_lon = 51.0447, -114.0719

print("Calgary's coordinates:")
print(f"Decimal Degrees (DD): {calgary_lat:.6f}°N, {abs(calgary_lon):.6f}°W")

# Convert DD to DMS
lat_d, lat_m, lat_s = dd_to_dms(calgary_lat)
lon_d, lon_m, lon_s = dd_to_dms(abs(calgary_lon))
print(f"Degrees, Minutes, Seconds (DMS): {lat_d}°{lat_m}'{lat_s:.4f}\"N, {lon_d}°{lon_m}'{lon_s:.4f}\"W")

# Convert DD to DMM
lat_d, lat_m = dd_to_dmm(calgary_lat)
lon_d, lon_m = dd_to_dmm(abs(calgary_lon))
print(f"Degrees and Decimal Minutes (DMM): {lat_d}°{lat_m:.4f}'N, {lon_d}°{lon_m:.4f}'W")

# Demonstrate conversion from DMS to DD
print("\nDemonstrating conversion from DMS to DD:")
calc_dd_lat = dms_to_dd(lat_d, lat_m, lat_s)
calc_dd_lon = -dms_to_dd(lon_d, lon_m, lon_s)  # Negative for western longitude
print(f"Calculated DD from DMS: {calc_dd_lat:.6f}°N, {calc_dd_lon:.6f}°W")

# Calculate the difference to show precision
lat_diff = abs(calgary_lat - calc_dd_lat)
lon_diff = abs(calgary_lon - calc_dd_lon)
print(f"Difference in latitude: {lat_diff:.9f}°")
print(f"Difference in longitude: {lon_diff:.9f}°")

# Demonstrate conversion from DD to DMS
print("\nDemonstrating conversion from DD to DMS:")
lat_d, lat_m, lat_s = dd_to_dms(calgary_lat)
lon_d, lon_m, lon_s = dd_to_dms(abs(calgary_lon))
print(f"Calculated DMS from DD: {lat_d}°{lat_m}'{lat_s:.4f}\"N, {lon_d}°{lon_m}'{lon_s:.4f}\"W")

Calgary's coordinates:
Decimal Degrees (DD): 51.044700°N, 114.071900°W
Degrees, Minutes, Seconds (DMS): 51°2'40.9200"N, 114°4'18.8400"W
Degrees and Decimal Minutes (DMM): 51°2.6820'N, 114°4.3140'W

Demonstrating conversion from DMS to DD:
Calculated DD from DMS: 51.056067°N, -114.077133°W
Difference in latitude: 0.011366667°
Difference in longitude: 0.005233333°

Demonstrating conversion from DD to DMS:
Calculated DMS from DD: 51°2'40.9200"N, 114°4'18.8400"W
